modules/gapi/include/opencv2/gapi/media.hpp

```diff
@@ -0,0 +1,72 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+#ifndef OPENCV_GAPI_MEDIA_HPP
+#define OPENCV_GAPI_MEDIA_HPP
+
+#include <memory>     // unique_ptr<>, shared_ptr<>
+#include <array>      // array<>
+#include <functional> // function<>
+#include <utility>    // forward<>()
+
+#include <opencv2/gapi/gframe.hpp>
+
+namespace cv {
+
+class GAPI_EXPORTS MediaFrame {
+public:
+    enum class Access { R, W };
+    class IAdapter;
+    class View;
+    using AdapterPtr = std::unique_ptr<IAdapter>;
+
+    MediaFrame();
+    explicit MediaFrame(AdapterPtr &&);
+    template<class T, class... Args> static cv::MediaFrame Create(Args&&...);
+
+    View access(Access);
+    cv::GFrameDesc desc() const;
+
+private:
+    struct Priv;
+    std::shared_ptr<Priv> m;
+};
+
+template<class T, class... Args>
+inline cv::MediaFrame cv::MediaFrame::Create(Args&&... args) {
+    std::unique_ptr<T> ptr(new T(std::forward<Args>(args)...));
+    return cv::MediaFrame(std::move(ptr));
+}
+
+class GAPI_EXPORTS MediaFrame::View final {
+public:
+    static constexpr const size_t MAX_PLANES = 4;
+    using Ptrs     = std::array<void*, MAX_PLANES>;
+    using Strides  = std::array<std::size_t, MAX_PLANES>; // in bytes
+    using Callback = std::function<void()>;
+
+    View(Ptrs&& ptrs, Strides&& strs, Callback &&cb);
+    View(const View&) = delete;
+    View(View&&) = default;
+    ~View();
+
+    std::array<void*, MAX_PLANES> ptr;
+    std::array<std::size_t, MAX_PLANES> stride;
+
+private:
+    Callback m_cb;
+};
+
+class GAPI_EXPORTS MediaFrame::IAdapter {
+public:
+    virtual ~IAdapter() = 0;
```

Comment:
This = 0 looks misleading since ~IAdapter is implemented. So it simply can be = default

Reply:
This is never misleading as it was quite common in pre-Cxx11 era.

My main concern about `= default` is that when a `media.hpp` is included in the user application, its destructor will be generated by the user's compiler (different from ours)? I don't trust much to this :)

Reply:
So my point here is that `= 0` usually says that user *must* implement this method in a derived class, but actually it turns out that there is an implementation in a base class and user can utilize this base class implementation, that's what I meant by misleading. You can ask your compiler to generate a destructor by putting `= default` in a destructor definition at .cpp file